File: ts/src/nougen_shards/graph.ts
/**
 * Graph Memory: link shards (fixes, files, commands, decisions) into a latent mesh.
 * (TS mimic of graph.py)
 *
 * Edges live in a dedicated graph.db inside the vault (honors NOUGEN_VAULT_DIR via
 * core.GLOBAL_DIR). Nodes are identified by file_hash, not the per-DB autoincrement
 * `id` (the 9-DB cluster gives each database its own id sequence, so id alone is not
 * unique across the mesh; file_hash is, since capture() dedups globally). The public
 * API takes shard ids for ergonomics, with a db_index (recall's `_db_index`) to say
 * which database the id lives in (default 1, the common single-DB case).
 */
import { existsSync, mkdirSync } from "node:fs";
import * as path from "node:path";
import { createDatabase, type DatabaseSync } from "./_db.js";
import * as core from "./core.js";
import type { Shard } from "./core.js";

/** Path to the graph edge store (alongside the shard cluster in the vault). */
export function get_graph_db_path(): string {
  mkdirSync(core.GLOBAL_DIR, { recursive: true });
  return path.join(core.GLOBAL_DIR, "graph.db");
}

function get_graph_connection(): DatabaseSync {
  return createDatabase(get_graph_db_path());
}

/** Initialize the shard_edges table and lookup indexes. */
export function init_graph_db(): void {
  const conn = get_graph_connection();
  conn.exec(`
        CREATE TABLE IF NOT EXISTS shard_edges (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            src_hash TEXT NOT NULL,
            dst_hash TEXT NOT NULL,
            relation TEXT NOT NULL DEFAULT 'relates',
            created_at TEXT NOT NULL,
            UNIQUE(src_hash, dst_hash, relation)
        )
    `);
  conn.exec("CREATE INDEX IF NOT EXISTS idx_edges_src ON shard_edges(src_hash)");
  conn.exec("CREATE INDEX IF NOT EXISTS idx_edges_dst ON shard_edges(dst_hash)");
  conn.close();
}

/** Resolve (id, db_index) -> file_hash, the global node identity. */
function _hash_for(shard_id: number, db_index: number): string | null {
  const shard = core.get_shard_by_id(shard_id, db_index);
  return shard ? (shard.file_hash as string) : null;
}

/** Resolve a file_hash back to its shard dict by scanning the cluster. */
function _shard_for_hash(file_hash: string): Shard | null {
  for (let i = 1; i <= core.MAX_DB_COUNT; i++) {
    if (!existsSync(core.get_db_path(i))) {
      continue;
    }
    const conn = core.get_connection(i);
    try {
      const row = conn.prepare("SELECT * FROM shards WHERE file_hash = ?").get(file_hash) as Shard | undefined;
      if (row) {
        const item: Shard = { ...row };
        item._db_index = i;
        return item;
      }
    } catch {
      /* table may not exist yet */
    } finally {
      conn.close();
    }
  }
  return null;
}

/**
 * Create an edge src -> dst labelled `relation` (e.g. 'fixes', 'touches',
 * 'caused_by', 'relates'). Both shards must exist. Idempotent on
 * (src_hash, dst_hash, relation). bidirectional=true also stores dst -> src.
 * Returns true if at least one new edge was written.
 */
export function link_shards(
  src_id: number,
  dst_id: number,
  relation: string = "relates",
  src_db: number = 1,
  dst_db: number = 1,
  bidirectional: boolean = false,
): boolean {
  const src_hash = _hash_for(src_id, src_db);
  const dst_hash = _hash_for(dst_id, dst_db);
  if (!src_hash || !dst_hash || src_hash === dst_hash) {
    return false;
  }

  init_graph_db();
  const timestamp = new Date().toISOString();
  const conn = get_graph_connection();
  try {
    let written = conn
      .prepare(
        "INSERT OR IGNORE INTO shard_edges (src_hash, dst_hash, relation, created_at) VALUES (?, ?, ?, ?)",
      )
      .run(src_hash, dst_hash, relation, timestamp).changes;
    if (bidirectional) {
      written += conn
        .prepare(
          "INSERT OR IGNORE INTO shard_edges (src_hash, dst_hash, relation, created_at) VALUES (?, ?, ?, ?)",
        )
        .run(dst_hash, src_hash, relation, timestamp).changes;
    }
    return Number(written) > 0;
  } finally {
    conn.close();
  }
}

/**
 * Return shards connected to (shard_id, db_index) — undirected: follows edges in
 * either direction — optionally filtered to a single relation. Each result is the
 * neighbour shard dict enriched with `relation` and `direction` ('out'/'in').
 */
export function related_shards(
  shard_id: number,
  db_index: number = 1,
  relation: string | null = null,
  limit: number = 10,
): Shard[] {
  const node_hash = _hash_for(shard_id, db_index);
  if (!node_hash || !existsSync(get_graph_db_path())) {
    return [];
  }

  const conn = get_graph_connection();
  let out_rows: Array<{ nhash: string; relation: string }>;
  let in_rows: Array<{ nhash: string; relation: string }>;
  try {
    const rel_sql = relation ? " AND relation = ?" : "";
    const out_params = relation ? [node_hash, relation] : [node_hash];
    out_rows = conn
      .prepare(`SELECT dst_hash AS nhash, relation FROM shard_edges WHERE src_hash = ?${rel_sql}`)
      .all(...out_params) as Array<{ nhash: string; relation: string }>;
    const in_params = relation ? [node_hash, relation] : [node_hash];
    in_rows = conn
      .prepare(`SELECT src_hash AS nhash, relation FROM shard_edges WHERE dst_hash = ?${rel_sql}`)
      .all(...in_params) as Array<{ nhash: string; relation: string }>;
  } finally {
    conn.close();
  }

  const neighbours: Shard[] = [];
  const seen = new Set<string>();
  const tagged = [
    ...out_rows.map((r) => ({ row: r, direction: "out" })),
    ...in_rows.map((r) => ({ row: r, direction: "in" })),
  ];
  for (const { row, direction } of tagged) {
    const key = `${row.nhash}|${row.relation}|${direction}`;
    if (seen.has(key)) {
      continue;
    }
    seen.add(key);
    const shard = _shard_for_hash(row.nhash);
    if (shard) {
      shard.relation = row.relation;
      shard.direction = direction;
      neighbours.push(shard);
    }
    if (neighbours.length >= limit) {
      break;
    }
  }
  return neighbours;
}

/** Total number of edges in the mesh (0 if the graph store is absent). */
export function edge_count(): number {
  if (!existsSync(get_graph_db_path())) {
    return 0;
  }
  const conn = get_graph_connection();
  try {
    return (conn.prepare("SELECT COUNT(*) AS c FROM shard_edges").get() as { c: number }).c;
  } catch {
    return 0;
  } finally {
    conn.close();
  }
}

File: src/nougen_shards/graph.py
"""
Graph Memory: link shards (fixes, files, commands, decisions) into a latent mesh.

Edges live in a dedicated graph.db inside the vault (honors NOUGEN_VAULT_DIR via
core.GLOBAL_DIR), so a fix can point at the file it touched, a command at the
decision that prompted it, and recall can walk those links.

Nodes are identified by file_hash, not by the per-DB autoincrement `id`: the
9-DB cluster gives each database its own id sequence, so id alone is not unique
across the mesh, but file_hash is (capture() dedups globally before writing).
The public API still takes shard ids for ergonomics, with a db_index to say
which database the id lives in (default 1, the common single-DB case; recall
results carry it as `_db_index`).
"""
import sqlite3
from datetime import datetime, timezone
from typing import List, Dict, Optional, Tuple

from . import core


def get_graph_db_path():
    """Path to the graph edge store (alongside the shard cluster in the vault)."""
    core.GLOBAL_DIR.mkdir(parents=True, exist_ok=True)
    return core.GLOBAL_DIR / "graph.db"


def get_graph_connection():
    """SQLite connection to the graph store, WAL-enabled (Module 19)."""
    conn = sqlite3.connect(str(get_graph_db_path()), timeout=10.0)
    conn.execute("PRAGMA journal_mode=WAL;")
    conn.row_factory = sqlite3.Row
    return conn


def init_graph_db():
    """Initialize the shard_edges table and lookup indexes."""
    conn = get_graph_connection()
    conn.execute("""
        CREATE TABLE IF NOT EXISTS shard_edges (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            src_hash TEXT NOT NULL,
            dst_hash TEXT NOT NULL,
            relation TEXT NOT NULL DEFAULT 'relates',
            created_at TEXT NOT NULL,
            UNIQUE(src_hash, dst_hash, relation)
        )
    """)
    conn.execute("CREATE INDEX IF NOT EXISTS idx_edges_src ON shard_edges(src_hash)")
    conn.execute("CREATE INDEX IF NOT EXISTS idx_edges_dst ON shard_edges(dst_hash)")
    conn.commit()
    conn.close()


def _hash_for(shard_id: int, db_index: int) -> Optional[str]:
    """Resolve (id, db_index) -> file_hash, the global node identity."""
    shard = core.get_shard_by_id(shard_id, db_index)
    return shard["file_hash"] if shard else None


def _shard_for_hash(file_hash: str) -> Optional[Dict]:
    """Resolve a file_hash back to its shard dict by scanning the cluster."""
    return _shards_for_hashes([file_hash]).get(file_hash)


def _shards_for_hashes(file_hashes: List[str]) -> Dict[str, Dict]:
    """Resolve many file_hashes to shard dicts with one query per DB (lowest db_index wins).

    Avoids the per-hash cluster rescan: a single pass over the (up to) 9 DBs
    resolves every requested hash, so callers pay at most one connection per DB
    instead of one per (hash, DB) pair.
    """
    remaining = list(dict.fromkeys(file_hashes))  # de-dup, preserve order
    resolved: Dict[str, Dict] = {}
    for i in range(1, core.MAX_DB_COUNT + 1):
        if not remaining:
            break
        if not core.get_db_path(i).exists():
            continue
        conn = core.get_connection(i)
        try:
            placeholders = ",".join("?" * len(remaining))
            rows = conn.execute(
                f"SELECT * FROM shards WHERE file_hash IN ({placeholders})", remaining
            ).fetchall()
            for row in rows:
                item = dict(row)
                fhash = item["file_hash"]
                if fhash not in resolved:  # first (lowest-index) DB wins
                    item["_db_index"] = i
                    resolved[fhash] = item
            remaining = [h for h in remaining if h not in resolved]
        except sqlite3.OperationalError:
            pass
        finally:
            conn.close()
    return resolved


def link_shards(src_id: int, dst_id: int, relation: str = "relates",
                src_db: int = 1, dst_db: int = 1, bidirectional: bool = False) -> bool:
    """
    Create an edge src -> dst labelled `relation` (e.g. 'fixes', 'touches',
    'caused_by', 'relates'). Both shards must exist. Idempotent on
    (src_hash, dst_hash, relation). bidirectional=True also stores dst -> src.
    Returns True if at least one new edge was written.
    """
    src_hash = _hash_for(src_id, src_db)
    dst_hash = _hash_for(dst_id, dst_db)
    if not src_hash or not dst_hash or src_hash == dst_hash:
        return False

    init_graph_db()
    timestamp = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
    conn = get_graph_connection()
    try:
        written = conn.execute(
            "INSERT OR IGNORE INTO shard_edges (src_hash, dst_hash, relation, created_at) "
            "VALUES (?, ?, ?, ?)", (src_hash, dst_hash, relation, timestamp)).rowcount
        if bidirectional:
            written += conn.execute(
                "INSERT OR IGNORE INTO shard_edges (src_hash, dst_hash, relation, created_at) "
                "VALUES (?, ?, ?, ?)", (dst_hash, src_hash, relation, timestamp)).rowcount
        conn.commit()
        return written > 0
    finally:
        conn.close()


def related_shards(shard_id: int, db_index: int = 1, relation: Optional[str] = None,
                   limit: int = 10) -> List[Dict]:
    """
    Return shards connected to (shard_id, db_index) — undirected: follows edges in
    either direction — optionally filtered to a single relation. Each result is the
    neighbour shard dict enriched with 'relation' and 'direction' ('out'/'in').
    """
    node_hash = _hash_for(shard_id, db_index)
    if not node_hash or not get_graph_db_path().exists():
        return []

    conn = get_graph_connection()
    try:
        rel_sql = " AND relation = ?" if relation else ""
        out_params: list = [node_hash] + ([relation] if relation else [])
        out_rows = conn.execute(
            f"SELECT dst_hash AS nhash, relation FROM shard_edges WHERE src_hash = ?{rel_sql}",
            out_params).fetchall()
        in_params: list = [node_hash] + ([relation] if relation else [])
        in_rows = conn.execute(
            f"SELECT src_hash AS nhash, relation FROM shard_edges WHERE dst_hash = ?{rel_sql}",
            in_params).fetchall()
    finally:
        conn.close()

    ordered = [(r, "out") for r in out_rows] + [(r, "in") for r in in_rows]
    hash_map = _shards_for_hashes([row["nhash"] for row, _ in ordered])

    neighbours: List[Dict] = []
    seen = set()
    for row, direction in ordered:
        key = (row["nhash"], row["relation"], direction)
        if key in seen:
            continue
        seen.add(key)
        shard = hash_map.get(row["nhash"])
        if shard:
            shard = dict(shard)  # copy: same hash may recur with a different relation/direction
            shard["relation"] = row["relation"]
            shard["direction"] = direction
            neighbours.append(shard)
        if len(neighbours) >= limit:
            break
    return neighbours


def edge_count() -> int:
    """Total number of edges in the mesh (0 if the graph store is absent)."""
    if not get_graph_db_path().exists():
        return 0
    conn = get_graph_connection()
    try:
        return conn.execute("SELECT COUNT(*) FROM shard_edges").fetchone()[0]
    except sqlite3.Error:
        return 0
    finally:
        conn.close()
